README.md
# node-statenspersonadressregister
A library that can be used to query to SPAR (https://www.statenspersonadressregister.se/).
SPAR is a service that is provided by the Swedish government for access to data about Swedish citizens.
Information on how to gain access to this service can be read on the SPAR website.

## Requirements
Make sure you have your certificate from Steria and the account information you need to access the test servers

## Usage
Make sure your project depend on `node-statenspersonadressregister` and then:

```
import SPAR from 'spar';
const SPAR_OPTIONS = {
  certPath: '', // file path to the certificate from Steria [cert.p12]
  passphrase: '', // The passphrase for the certificate
  kundNr: '', // From SPAR
  orgNr: '', // From SPAR,
  slutAnvandarId: '', // From SPAR,
  baseUrl: '', // url to SPAR
};

const spar = new SPAR(SPAR_OPTIONS);
```

### Personsök

```
spar.personSok(PERSONNUMMER).then((result) => {
  // result is an array of results, should only be 1 entry
}, (error) => {
  // something went wrong, error has more info
});
```


## More information
https://www.statenspersonadressregister.se/download/18.41febba7149eb27c07a50fb/1447748382337/SPAR_Systemgränssnitt_ver_1-74.pdf


CHANGELOG.md
## 3.0
* Remove cURL and use request instead
* Remove pem certificate, use pfx and passphrase
* Use babel env to build for node 4

## 2.0
* Use SPAR xml version 20160213
* Add support for jest testing


src/__test__/XMLParser.test.js
import XMLParser from '../XMLParser';

const SPAR_XML_RESPONSE_2016 = `
<?xml version="1.0" encoding="utf-8" ?>
<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/"><SOAP-ENV:Header/>
    <SOAP-ENV:Body>
        <SPARPersonsokningSvar xmlns="http://skatteverket.se/spar/instans/1.0">
            <spako:PersonsokningFraga xmlns:spako="http://skatteverket.se/spar/komponent/1.0">
                <spako:PersonId>
                    <spako:FysiskPersonId>198904079558</spako:FysiskPersonId>
                </spako:PersonId>
            </spako:PersonsokningFraga>
            <spako:PersonsokningSvarsPost xmlns:spako="http://skatteverket.se/spar/komponent/1.0">
                <spako:PersonId>
                    <spako:FysiskPersonId>198904079558</spako:FysiskPersonId>
                </spako:PersonId>
                <spako:Sekretessmarkering>N</spako:Sekretessmarkering>
                <spako:SenasteAndringSPAR>2016-08-30</spako:SenasteAndringSPAR>
                <spako:Persondetaljer>
                    <spako:DatumFrom>2009-09-10</spako:DatumFrom>
                    <spako:DatumTill>9999-12-31</spako:DatumTill>
                    <spako:Fornamn>Fornamn Tilltalsnamn</spako:Fornamn>
                    <spako:Tilltalsnamn>20</spako:Tilltalsnamn>
                    <spako:Efternamn>Efternamn</spako:Efternamn>
                    <spako:Fodelsetid>1989-04-07</spako:Fodelsetid>
                    <spako:Kon>M</spako:Kon>
                </spako:Persondetaljer>
                <spako:Adresser>
                    <spako:Folkbokforingsadress>
                        <spako:DatumFrom>2016-08-30</spako:DatumFrom>
                        <spako:DatumTill>9999-12-31</spako:DatumTill>
                        <spako:Utdelningsadress2>GATAN 123 LGH 1234</spako:Utdelningsadress2>
                        <spako:PostNr>59242</spako:PostNr>
                        <spako:Postort>Postorten</spako:Postort>
                        <spako:FolkbokfordLanKod>06</spako:FolkbokfordLanKod>
                        <spako:FolkbokfordKommunKod>23</spako:FolkbokfordKommunKod>
                        <spako:Folkbokforingsdatum>2014-08-25</spako:Folkbokforingsdatum>
                        <spako:DistriktKod>112042</spako:DistriktKod>
                    </spako:Folkbokforingsadress>
                    <spako:Folkbokforingsadress>
                        <spako:DatumFrom>2016-01-05</spako:DatumFrom>
                        <spako:DatumTill>2016-08-30</spako:DatumTill>
                        <spako:Utdelningsadress2>ANDRA GATAN 567 2TR</spako:Utdelningsadress2>
                        <spako:PostNr>18204</spako:PostNr>
                        <spako:Postort>Staden</spako:Postort>
                        <spako:FolkbokfordLanKod>04</spako:FolkbokfordLanKod>
                        <spako:FolkbokfordKommunKod>17</spako:FolkbokfordKommunKod>
                        <spako:Folkbokforingsdatum>2014-08-25</spako:Folkbokforingsdatum>
                        <spako:DistriktKod>39203</spako:DistriktKod>
                    </spako:Folkbokforingsadress>
                    <spako:Folkbokforingsadress>
                        <spako:DatumFrom>2014-08-26</spako:DatumFrom>
                        <spako:DatumTill>2016-01-05</spako:DatumTill>
                        <spako:Utdelningsadress2>VÄGEN 2 2TR</spako:Utdelningsadress2>
                        <spako:PostNr>24813</spako:PostNr>
                        <spako:Postort>PLATSEN</spako:Postort>
                        <spako:FolkbokfordLanKod>04</spako:FolkbokfordLanKod>
                        <spako:FolkbokfordKommunKod>11</spako:FolkbokfordKommunKod>
                        <spako:FolkbokfordForsamlingKod>02</spako:FolkbokfordForsamlingKod>
                        <spako:Folkbokforingsdatum>2014-08-25</spako:Folkbokforingsdatum>
                    </spako:Folkbokforingsadress>
                    <spako:Folkbokforingsadress>
                        <spako:DatumFrom>2011-02-09</spako:DatumFrom>
                        <spako:DatumTill>2014-08-26</spako:DatumTill>
                        <spako:Utdelningsadress2>STIGEN 152 LGH 1612</spako:Utdelningsadress2>
                        <spako:PostNr>49252</spako:PostNr>
                        <spako:Postort>Byn</spako:Postort>
                        <spako:FolkbokfordLanKod>25</spako:FolkbokfordLanKod>
                        <spako:FolkbokfordKommunKod>98</spako:FolkbokfordKommunKod>
                        <spako:FolkbokfordForsamlingKod>12</spako:FolkbokfordForsamlingKod>
                        <spako:Folkbokforingsdatum>2011-01-31</spako:Folkbokforingsdatum>
                    </spako:Folkbokforingsadress>
                </spako:Adresser>
            </spako:PersonsokningSvarsPost>
        </SPARPersonsokningSvar>
    </SOAP-ENV:Body>
</SOAP-ENV:Envelope>
`;

describe('The parser can parse SPAR bodies', () => {
  it('hejsan', () => {
    const parser = new XMLParser();
    const answer = parser.parsePersonsokningResponse(SPAR_XML_RESPONSE_2016);

    expect(answer).toMatchSnapshot();

    expect(true).toBe(true);
  });
});


src/index.js
import fs from 'fs';
import request from 'request';

import XMLParser from './XMLParser';
import XMLFormatter from './XMLFormatter';

const DEFAULT_CONFIG = {
  certPath: '',
  passphrase: '',
  baseUrl: 'https://kt-ext-ws.statenspersonadressregister.se/spar-webservice/SPARPersonsokningService/20160213',
  kundNr: '',
  orgNr: '',
  slutAnvandarId: '',
  verbose: false,
};

class SPAR {
  constructor(config, { parser = null, formatter = null } = {}) {
    this.config = Object.assign({}, DEFAULT_CONFIG, config);
    this.certificate = fs.readFileSync(this.config.certPath);

    if (parser === null) {
      this.parser = new XMLParser();
    } else {
      this.parser = parser;
    }

    if (formatter === null) {
      this.formatter = new XMLFormatter();
    } else {
      this.formatter = formatter;
    }
  }

  personSok(fysiskPersonId) {
    const {
      kundNr,
      orgNr,
      slutAnvandarId,
    } = this.config;

    const searchEnvelope = this.formatter.formatPersonsokningFraga(
      fysiskPersonId, kundNr, orgNr, slutAnvandarId
    );

    return this.soapCall(searchEnvelope).then((body) => {
      return new Promise((resolve, reject) => {
        const answer = this.parser.parsePersonsokningResponse(body);
        if (answer.Undantag.length > 0) {
          reject(answer.Undantag);
        } else {
          resolve(answer.PersonsokningSvarsPost);
        }
      });
    });
  }

  soapCall(envelope) {
    const { verbose, baseUrl, passphrase } = this.config;

    const options = {
      url: baseUrl,
      agentOptions: {
        pfx: this.certificate,
        passphrase: passphrase,
      },
      contentType: 'text/xml; charset=utf-8',
      body: envelope,
    };

    return new Promise((resolve, reject) => {
      request.post(options, (err, res, body) => {
        if (err) {
          reject(err);
        } else {
          resolve(body);
        }
      });
    });
  }

  _validateConfig(/* config */) {
    return true;
  }
}

export default SPAR;


src/XMLFormatter.js
import xml from 'xml';
import moment from 'moment';

class XMLFormatter {
  formatPersonsokningFraga(fysiskPersonId, kundNr, orgNr, slutAnvandarId) {
    const searchData = {
      'spain:SPARPersonsokningFraga': [
        { '_attr': { 'xmlns:spain': 'http://skatteverket.se/spar/instans/1.0' } },
        this._formatSPARIdentity(kundNr, orgNr, slutAnvandarId),
        this._formatSPARPersonsokningFraga(fysiskPersonId),
      ],
    };

    return this._formatSPAREnvelope(searchData);
  }

  _formatSPARIdentity(kundNr, orgNr, slutAnvandarId) {
    return {
      'spako:IdentifieringsInformation': [
        { '_attr': {
          'xmlns:spako': 'http://skatteverket.se/spar/komponent/1.0',
        } },

        { 'spako:KundNrLeveransMottagare': kundNr },
        { 'spako:KundNrSlutkund': kundNr },
        { 'spako:OrgNrSlutkund': orgNr },
        { 'spako:SlutAnvandarId': slutAnvandarId },
        { 'spako:Tidsstampel': moment().format('YYYY-MM-DDTHH:mm:ss.SSSS') },
      ],
    };
  }

  _formatSPARPersonsokningFraga(idNum) {
    return {
      'spako:PersonsokningFraga': [
        { '_attr': {
          'xmlns:spako': 'http://skatteverket.se/spar/komponent/1.0',
        } },

        { 'spako:PersonId': [
          { 'spako:FysiskPersonId': idNum },
        ]},
      ],
    };
  }

  _formatSPAREnvelope(soapBody) {
    return xml({
      'SOAP-ENV:Envelope': [
        { '_attr': {
          'xmlns:SOAP-ENV': 'http://schemas.xmlsoap.org/soap/envelope/',
          'xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance',
          'SOAP-ENV:encodingStyle': 'http://schemas.xmlsoap.org/soap/encoding/',
        } },
        { 'SOAP-ENV:Body': [soapBody] },
      ],
    });
  }
}

export default XMLFormatter;


src/XMLParser.js
import { DOMParser } from 'xmldom';
import xpath from 'xpath';

const getDate = (dateStr) => {
  if (dateStr === '9999-12-31') {
    return null;
  }

  return dateStr;
};

/**
 * Given an array of Objects, get the current.
 * It's the one with no `DatumTill`.
 * @return {Object}
 */
const getCurrent = (objects) => {
  let current = null;
  objects.some((object) => {
    if (object.DatumTill === null) {
      current = object;

      return true;
    }

    return false;
  });

  return current;
};

/**
 * @param  {String} utdelningsadress
 * @return {{ utdelningsadress, lagenhet }}
 */
const splitUtdelningsadress = (inUtdelningsadress) => { // esl
  const LGH_REGEX = /\s+(LGH|LÄG)\s+(\d+)/;
  const TR_REGEX = /\s+(\d+)TR/;

  let Utdelningsadress = inUtdelningsadress;
  let Lagenhet;
  let Trappor;

  if (LGH_REGEX.test(Utdelningsadress)) {
    // Get the LGH number
    Lagenhet = Utdelningsadress.match(LGH_REGEX)[2];
    // Get everything except the LGH part
    Utdelningsadress = Utdelningsadress.replace(LGH_REGEX, '');
  }

  if (TR_REGEX.test(Utdelningsadress)) {
    // Get the trappor number
    Trappor = Utdelningsadress.match(TR_REGEX)[1];
    // Remove the trappor part
    Utdelningsadress = Utdelningsadress.replace(TR_REGEX, '');
  }

  return { Utdelningsadress, Lagenhet, Trappor };
};

class XMLParser {
  constructor(namespaces = {
    'spain': 'http://skatteverket.se/spar/instans/1.0',
    'spako': 'http://skatteverket.se/spar/komponent/1.0',
    'SOAP-ENV': 'http://schemas.xmlsoap.org/soap/envelope/',
  }) {
    this.select = xpath.useNamespaces(namespaces);
  }

  parsePersonsokningResponse(response) {
    const dom = (new DOMParser()).parseFromString(response);

    const UndantagNodes = this.select('//spako:Undantag', dom);
    const PersonsokningSvarsPostNodes = this.select('//spako:PersonsokningSvarsPost', dom);

    return {
      Undantag: UndantagNodes
        .map((UndantagNode) => this.parseUndantag(UndantagNode)),
      PersonsokningSvarsPost: PersonsokningSvarsPostNodes
        .map((node) => this.parsePersonsokningSvarsPost(node)),
    };
  }

  parsePersonsokningSvarsPost(personsokningSvarsPostNode) {
    const getString = this._getStringCreator(personsokningSvarsPostNode);

    const PersonId = this.select('spako:PersonId', personsokningSvarsPostNode)
      .map((PersonIdNode) => this.parsePersonId(PersonIdNode));

    const Persondetaljer = this.select('spako:Persondetaljer', personsokningSvarsPostNode)
      .map((persondetaljer) => this.parsePersondetaljer(persondetaljer));
    const AktuellPersondetaljer = getCurrent(Persondetaljer);

    const Folkbokforingsadresser = this.select('//spako:Folkbokforingsadress', personsokningSvarsPostNode)
      .map((address) => this.parseFolkbokforingsadress(address));
    const AktuellFolkbokforingsadress = getCurrent(Folkbokforingsadresser);

    return {
      PersonId,

      Sekretessmarkering: getString('Sekretessmarkering'),
      SekretessAndringsdatum: getDate(getString('SekretessAndringsdatum')),
      SenasteAndringSPAR: getDate(getString('SenasteAndringSPAR')),

      Persondetaljer,
      AktuellPersondetaljer,
      Folkbokforingsadresser,
      AktuellFolkbokforingsadress,
    };
  }

  parsePersonId(node) {
    const getString = this._getStringCreator(node);

    return {
      FysiskPersonId: getString('FysiskPersonId'),
    };
  }

  parsePersondetaljer(node) {
    const getString = this._getStringCreator(node);
    const fornamns = getString('Fornamn').split(' ');

    // Tilltalsnamn is saved in SPAR as a 2 digit code
    // if the second number is 0 then the first number
    // defines which of the first names are the Tilltalsnamn
    // (index starts at 1)
    let tilltalsnamn = '';
    const tilltalsnamnCode = getString('Tilltalsnamn');
    if (tilltalsnamnCode === '') {
      tilltalsnamn = fornamns[0];
    } else {
      const [a, b] =
        tilltalsnamnCode
        .split('') // Make the code an array
        .map((c) => parseInt(c, 10)) // convert the strings to numbers
        .map((i) => i - 1); // make the name index start at 0

      if (b === -1) {
        tilltalsnamn = fornamns[a];
      } else {
        tilltalsnamn = `${ fornamns[a] } ${ fornamns[b] }`;
      }
    }

    return {
      DatumFrom: getDate(getString('DatumFrom')),
      DatumTill: getDate(getString('DatumTill')),
      Fornamn: fornamns,
      Tilltalsnamn: tilltalsnamn,
      Efternamn: getString('Efternamn'),
      Aviseringsnamn: getString('Aviseringsnamn'),
      HanvisningspersonNrByttTill: getString('HanvisningspersonNrByttTill'),
      HanvisningspersonNrByttFran: getString('HanvisningspersonNrByttFran'),
      Avregistreringsdatum: getDate(getString('Avregistreringsdatum')),
      AvregistreringsorsakKod: getString('AvregistreringsorsakKod'),
      Fodelsetid: getDate(getString('Fodelsetid')),
      Kon: getString('Kon'),
    };
  }

  parseFolkbokforingsadress(node) {
    const getString = this._getStringCreator(node);

    return {
      DatumFrom: getDate(getString('DatumFrom')),
      DatumTill: getDate(getString('DatumTill')),
      CareOf: getString('CareOf'),
      DistriktKod: getString('DistriktKod'),
      FolkbokfordKommunKod: getString('FolkbokfordKommunKod'),
      FolkbokfordLanKod: getString('FolkbokfordLanKod'),
      Folkbokforingsdatum: getDate(getString('Folkbokforingsdatum')),
      Utdelningsadress1: splitUtdelningsadress(getString('Utdelningsadress1')),
      Utdelningsadress2: splitUtdelningsadress(getString('Utdelningsadress2')),
      PostNr: getString('PostNr'),
      Postort: getString('Postort'),
    };
  }

  parseUtlandsadress(node) {
    const getString = this._getStringCreator(node);

    return {
      Utdelningsadress1: splitUtdelningsadress(getString('Utdelningsadress1')),
      Utdelningsadress2: splitUtdelningsadress(getString('Utdelningsadress2')),
      Utdelningsadress3: splitUtdelningsadress(getString('Utdelningsadress3')),
      Land: getString('Land'),
    };
  }

  parseUndantag(node) {
    const getString = this._getStringCreator(node);

    return {
      Kod: getString('Kod'),
      Beskrivning: getString('Beskrivning'),
    };
  }

  _getStringCreator(node, prefix = 'spako:') {
    return (key) => this.select(`${ prefix }${ key }/text()`, node).toString();
  }
}

export default XMLParser;
